### PR TITLE
[WIP] Elastic Transform Implementation

### PR DESCRIPTION
**Pull request by IssamLaradji**
I would like to add a differentiable elastic transform function. This is my first attempt. 

The function is based on 
[Simard2003]: Simard, Steinkraus and Platt, "Best Practices for
                      Convolutional Neural Networks applied to Visual Document Analysis", in
                      Proc. of the International Conference on Document Analysis and
                      Recognition, 2003.

### CODE REVIEW

**Comment by edgarriba** [comment]
ping @IssamLaradji

File: kornia/geometry/transform/elastic_transform.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
+"""Elastic Transform
+"""
+
+# Authors: Issam H. Laradji <issam.laradji@gmail.com>
```
[the remark this answers is not in the document]
Comment (IssamLaradji) [reply kornia/geometry/transform/elastic_transform.py:4]: agreed!

File: kornia/geometry/transform/elastic_transform.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+import torch 
+import math
+import kornia
+
+import torch.nn.functional as F
+
+
+def elastic_transform_2d(img, kernel_size=(3,3), sigma=(4,4), alpha=(32,32), random_seed=None):
```
[the remark this answers is not in the document]
Comment (IssamLaradji) [reply kornia/geometry/transform/elastic_transform.py:13]: thanks adding that

File: kornia/geometry/transform/elastic_transform.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    r"""Applies elastic transform of images as described in [Simard2003]_.
+
+    Args:
+        img (torch.Tensor): input image.
```
[the remark this answers is not in the document]
Comment (IssamLaradji) [reply kornia/geometry/transform/elastic_transform.py:17]: good point

File: kornia/geometry/transform/elastic_transform.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        kernel_size (Tuple[int, int]): the size of the Gaussian kernel. Default:(3,3).
+        sigma (Tuple[float, float]): the standard deviation of the Gaussian in the y and x directions, respecitvely. 
+                                     Larger sigma results in smaller pixel displacements. Default:(4,4).
+        alpha (Tuple[float, float]):  the scaling factor that controls the intensity of the deformation
+                                  in the y and x directions, respectively. Default:32.
```
[the remark this answers is not in the document]
Comment (IssamLaradji) [reply kornia/geometry/transform/elastic_transform.py:22]: done

File: kornia/geometry/transform/elastic_transform.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    H, W = img.shape[-2:]
+    
+    # Get Gaussian kernel for 'y' and 'x' displacement
+    kernel_y = kornia.filters.get_gaussian_kernel2d(kernel_size, (sigma[0], sigma[0]))[None, None]
+    kernel_x = kornia.filters.get_gaussian_kernel2d(kernel_size, (sigma[1], sigma[1]))[None, None]
+
+    # Convolve over a random displacement matrix and scale them with 'alpha'
+    displacement = torch.rand(2, H, W, generator=generator)*0.1
+    disp_y = F.conv2d(displacement[0][None, None], weight=kernel_y, padding=kernel_size[0]//2) * alpha[0]
```
[the remark this answers is not in the document]
Comment (IssamLaradji) [reply kornia/geometry/transform/elastic_transform.py:46]: can you tell me the difference between conv2d and filter2d?

File: kornia/geometry/transform/elastic_transform.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    disp_x = F.conv2d(displacement[1][None, None], weight=kernel_x, padding=kernel_size[0]//2) * alpha[1]
+
+    # stack and normalize displacement
+    disp = torch.stack([disp_y, disp_x], dim=4).squeeze(0) 
+    
+    # Warp image based on displacement matrix
+    grid = kornia.utils.create_meshgrid(H, W)
```
[the remark this answers is not in the document]
Comment (IssamLaradji) [reply kornia/geometry/transform/elastic_transform.py:53]: fixed

**Comment by IssamLaradji** [comment]
@edgarriba  I have updated the `elastic_transform_2d` function based on your comments. What do you think?

I will work on the test file now

**Comment by edgarriba** [comment]
I'm gonna finish this PR. /cc @shijianjian @ducha-aiki @anguelos where should we place such kind of transformation ? as for now it's proposed to be included in `kornia.geometry.transform` - to me seems fine, but open to other suggestions.

**Comment by anguelos** [comment]
I believe kornia.geometry should not contain functions that are non-deterministic.
Actually random sampling in any way, is by definition not functional, as the RNG changes.
Shouldn't this go in augmentations?
We could maybe make it functional if we pass a noise image as a parameter.

File: kornia/geometry/transform/elastic_transform.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    n, c, h, w = tensor.shape
+    
+    # Get Gaussian kernel for 'y' and 'x' displacement
+    kernel_y = get_gaussian_kernel2d(kernel_size, (sigma[0], sigma[0]))[None]
+    kernel_x = get_gaussian_kernel2d(kernel_size, (sigma[1], sigma[1]))[None]
+
+    # Convolve over a random displacement matrix and scale them with 'alpha'
+    disp = torch.rand(n, 2, h, w, generator=generator).to(device=tensor.device)
+    disp[:,0] *= disp_scale[0]
+    disp[:,1] *= disp_scale[1]
```
Comment (edgarriba) [inline kornia/geometry/transform/elastic_transform.py:49]: @IssamLaradji where these disp_scale comes from ?

**Comment by edgarriba** [comment]
@anguelos makes sense. Will add `img_noise: torch.Tensor` parameter in the signatures expecting a `B2HW` noise image.

File: kornia/geometry/transform/elastic_transform.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+    disp_y = kornia.filters.filter2D(disp[:,[0]], kernel=kernel_y, border_type='reflect') * alpha[0]
+    disp_x = kornia.filters.filter2D(disp[:,[1]], kernel=kernel_x, border_type='reflect') * alpha[1]
+
+    # scale displacements
+    disp_y *= alpha[0]
+    disp_x *= alpha[1]
+
+    # stack and normalize displacement
+    disp = torch.cat([disp_y, disp_x], dim=1).squeeze(0).permute(0,2,3,1)
```
Comment (edgarriba) [inline kornia/geometry/transform/elastic_transform.py:59]: @IssamLaradji is this displacement assumed to be in pixel precision ?

File: kornia/geometry/transform/elastic_transform.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    disp = torch.rand(n, 2, h, w, generator=generator).to(device=tensor.device)
+    disp[:,0] *= disp_scale[0]
+    disp[:,1] *= disp_scale[1]
+
+    disp_y = kornia.filters.filter2D(disp[:,[0]], kernel=kernel_y, border_type='reflect') * alpha[0]
+    disp_x = kornia.filters.filter2D(disp[:,[1]], kernel=kernel_x, border_type='reflect') * alpha[1]
+
+    # scale displacements
+    disp_y *= alpha[0]
+    disp_x *= alpha[1]
```
Comment (edgarriba) [inline kornia/geometry/transform/elastic_transform.py:56]: @IssamLaradji why alpha is applied twice ? wouldn't be the same as to raise to the power of 2 ?

File: kornia/geometry/transform/elastic_transform.py (hunk 8 above)
Comment (IssamLaradji) [reply kornia/geometry/transform/elastic_transform.py:59]: Yes the displacement is in pixel precision

File: kornia/geometry/transform/elastic_transform.py (hunk 9 above)
Comment (IssamLaradji) [reply kornia/geometry/transform/elastic_transform.py:56]: ah the first one is a mistake, it should only be applied once

**Comment by IssamLaradji** [comment]
This code is outdated. The latest one is here:
https://github.com/IssamLaradji/elastic_transform/blob/master/elastic_transform.py

Do you want me to update this code to make it like the one in the link above?

**Pull request closed without merging**